### PR TITLE
k8s: Fix missing status subresource on CiliumNode

### LINKED ISSUE #13193: Cilium fails to update status on CiliumNode resource, leading to Cilium never becoming ready
This issue seems to only be observable on EKS environments. Because the issue is failing to update the CiliumNode status, likely Azure environments are broken as well. GKE is not affected because this only bug exists with IPAM modes that use the CiliumNode resource to hold IPAM information.

Cilium version: `master`
Steps to reproduce:
1) Follow EKS GSG guide
2) Deploy latest `master`.

I traced this down to this call here: https://github.com/cilium/cilium/blob/868feac8abcb70159f2a5709c5c36acbda14ac22/pkg/ipam/node.go#L716

I added an extra log statement:

```
diff --git a/pkg/ipam/node.go b/pkg/ipam/node.go
index ec000292a..0d96156cb 100644
--- a/pkg/ipam/node.go
+++ b/pkg/ipam/node.go
@@ -714,6 +714,9 @@ func (n *Node) syncToAPIServer() (err error) {
 
 		n.ops.PopulateStatusFields(node)
 		updatedNode, err = n.manager.k8sAPI.UpdateStatus(origNode, node)
+		if err != nil {
+			scopedLog.WithError(err).Warning("Failed to update status of node")
+		}
 		if updatedNode != nil && updatedNode.Name != "" {
 			node = updatedNode.DeepCopy()
 			if err == nil {
```

And that leads me to see the following log msgs:

```
[cilium-operator-58678745f4-nmftv] level=info msg="Synchronized ENI information" numInstances=4 numSecurityGroups=20 numSubnets=19 numVPCs=5 subsys=eni
[cilium-operator-58678745f4-nmftv] level=warning msg="Failed to update status of node" error="ciliumnodes.cilium.io \"ip-192-168-86-159.us-east-2.compute.internal\" not found" instanceID=i-0e986523decf7234c name=ip-192-168-86-159.us-east-2.compute.internal subsys=ipam
[cilium-operator-58678745f4-nmftv] level=warning msg="Failed to update status of node" error="ciliumnodes.cilium.io \"ip-192-168-5-163.us-east-2.compute.internal\" not found" instanceID=i-0b2545c92023c4eb8 name=ip-192-168-5-163.us-east-2.compute.internal subsys=ipam
[cilium-operator-58678745f4-nmftv] level=warning msg="Failed to update status of node" error="ciliumnodes.cilium.io \"ip-192-168-5-163.us-east-2.compute.internal\" not found" instanceID=i-0b2545c92023c4eb8 name=ip-192-168-5-163.us-east-2.compute.internal subsys=ipam
[cilium-operator-58678745f4-nmftv] level=warning msg="Failed to update status of node" error="ciliumnodes.cilium.io \"ip-192-168-86-159.us-east-2.compute.internal\" not found" instanceID=i-0e986523decf7234c name=ip-192-168-86-159.us-east-2.compute.internal subsys=ipam
```

Seems like we swallow the error a few lines down if a subsequent `Get` request for the CN resource succeeds.

### PR DESCRIPTION
**Pull request by christarazi**
This fixes the issue where we fail to update the status of the
CiliumNode resource, leading to failures in surfacing ENI or Azure IPAM
information.

Fixes: https://github.com/cilium/cilium/issues/13193

```release-note
Fix bug in EKS environments where Cilium agents never become ready due to the CiliumNode CRD disallowing updates
```

### CODE REVIEW

**Comment by christarazi** [comment]
test-me-please

File: pkg/k8s/apis/cilium.io/v2/types.go [hunk 1]
```diff
@@ -292,6 +292,7 @@ type CiliumEndpointList struct {
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 // +kubebuilder:resource:singular="ciliumnode",path="ciliumnodes",scope="Cluster",shortName={cn,ciliumn}
 // +kubebuilder:storageversion
+// +kubebuilder:subresource:status
```
Comment (sayboras) [inline pkg/k8s/apis/cilium.io/v2/types.go:295]: I believe that you already scanned once to make sure this is only resource missing `subresource:status` :)

**Review by sayboras — APPROVED** [review]
LGTM

**Review by aanm — CHANGES_REQUESTED** [review]
Please write a good release note for our users.

**Review by qmonnet — APPROVED** [review]
Thanks a lot for tracking this down!

**Review by aditighag — APPROVED** [review]

**Comment by joestringer** [comment]
> Please write a good release note for our users.

This is good practice in general, although it's more important if it actually affects an existing release. If I follow correctly, this is only affecting the master branch right now. @christarazi is that correct?

For the release note, I think the current PR title is fairly close, but it does not make it clear for instance what the observed error is (eg in Cilium logs) or the potential consequence. These would be helpful to assist users hitting an issue to suggest whether this bugfix is relevant to them or not.

**Comment by christarazi** [comment]
> > Please write a good release note for our users.
> 
> This is good practice in general, although it's more important if it actually affects an existing release. If I follow correctly, this is only affecting the master branch right now. @christarazi is that correct?
> 
> For the release note, I think the current PR title is fairly close, but it does not make it clear for instance what the observed error is (eg in Cilium logs) or the potential consequence. These would be helpful to assist users hitting an issue to suggest whether this bugfix is relevant to them or not. Feel free to ping me if you want to iterate on this and clarify the level of detail we would typically expect in a release note.

Yes that's correct. This is a regression on `master`. Will update the release note.

**Comment by christarazi** [comment]
test-me-please

**Comment by christarazi** [comment]
CI was failing because we needed 851a9ab. Unsure why we didn't see this behavior before with the missing status subresource.

**Comment by christarazi** [comment]
test-me-please

**Review by aanm — APPROVED** [review]

**Pull request merged**